rust/worker/src/system/types.rs

```diff
@@ -135,6 +138,121 @@ where
+    join_handle: Option<tokio::task::JoinHandle<()>>,
+    cancel: tokio_util::sync::CancellationToken,
+}
+
+#[derive(Clone)]
+pub(crate) struct Scheduler {
+    handles: Arc<Mutex<Vec<SchedulerHandle>>>,
+}
+
+impl Scheduler {
+    pub(crate) fn new() -> Scheduler {
+        Scheduler {
+            handles: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    pub(crate) fn schedule<C, M>(
+        &self,
+        sender: Sender<C>,
+        message: M,
+        duration: Duration,
+        ctx: &ComponentContext<C>,
+    ) where
+        C: Component + Handler<M>,
+        M: Debug + Send + 'static,
+    {
+        let cancel = ctx.cancellation_token.clone();
+        let handle = tokio::spawn(async move {
+            select! {
+                _ = cancel.cancelled() => {
+                    return;
+                }
+                _ = tokio::time::sleep(duration) => {
+                    match sender.send(message).await {
+                        Ok(_) => {
+                            return;
+                        },
+                        Err(e) => {
+                            // TODO: log error
+                            println!("Error: {:?}", e);
+                            return;
+                        }
+                    }
+                }
+            }
+        });
+        let handle = SchedulerHandle {
+            join_handle: Some(handle),
+            cancel: ctx.cancellation_token.clone(),
+        };
+        self.handles.lock().push(handle);
+    }
+
+    pub(crate) fn schedule_interval<C, M>(
+        &self,
+        sender: Sender<C>,
+        message: M,
+        duration: Duration,
+        ctx: &ComponentContext<C>,
+    ) where
+        C: Component + Handler<M>,
+        M: Debug + Send + Clone + 'static,
+    {
+        let cancel = ctx.cancellation_token.clone();
+        let handle = tokio::spawn(async move {
+            loop {
+                select! {
+                    _ = cancel.cancelled() => {
+                        return;
+                    }
+                    _ = tokio::time::sleep(duration) => {
+                        match sender.send(message.clone()).await {
+                            Ok(_) => {
+                                continue;
+                            },
+                            Err(e) => {
+                                // TODO: log error
+                                println!("Error: {:?}", e);
+                                continue;
+                            }
+                        }
+                    }
+                }
+            }
+        });
+        let handle = SchedulerHandle {
+            join_handle: Some(handle),
+            cancel: ctx.cancellation_token.clone(),
+        };
+        self.handles.lock().push(handle);
+    }
+
+    pub(crate) async fn join(&self) {
+        let mut handles = self.handles.lock();
+        for handle in handles.iter_mut() {
+            handle.cancel.cancel();
```

Comment:
I don't think join should cancel. Join is a way of saying "Wait until completed". I think Cancellation/Shutdown and Join are separate. Eg. Joining a thread or task doesn't kill that task